### PR TITLE
[WIP] CRM-16189, store credit card type for transaction

### PR DESCRIPTION
**Pull request by pradpnayak**
- [CRM-16189: Improve support for Accrual Method bookkeeping](https://issues.civicrm.org/jira/browse/CRM-16189)

### CODE REVIEW

File: CRM/Contribute/Form/Contribution.php [hunk 1]
```diff
@@ -1252,7 +1260,10 @@ protected function processCreditCard($submittedValues, $lineItem, $contactID) {
       // NOTE - I expect this is obsolete.
       $payment = Civi\Payment\System::singleton()->getByProcessor($this->_paymentProcessor);
       try {
-        $statuses = CRM_Contribute_BAO_Contribution::buildOptions('contribution_status_id');
+        $statuses = CRM_Core_PseudoConstant::get('CRM_Contribute_DAO_Contribution',
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/Contribution.php]: why did you make this change? My understanding is that @colemanw intends the buildOptions as the preferred method???

File: CRM/Contribute/Form/PaymentDetails.php [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+   */
+  public static function setDefaultValues(&$defaults, $contriId) {
+    //CRM-13420
+    if (empty($defaults['payment_instrument_id'])) {
+      $defaults['payment_instrument_id'] = key(CRM_Core_OptionGroup::values('payment_instrument', FALSE, FALSE, FALSE, 'AND is_default = 1'));
+    }
+
+    list($defaults['receive_date'], $defaults['receive_date_time']) = CRM_Utils_Date::setDateDefaults();
+
+    // CRM-4395
```
Comment (eileenmcnaughton) [inline CRM/Contribute/Form/PaymentDetails.php]: Are these really old CRMs still relevant?

File: templates/CRM/Contribute/Form/PaymentDetails.tpl [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+ | You should have received a copy of the GNU Affero General Public   |
+ | License and the CiviCRM Licensing Exception along                  |
+ | with this program; if not, contact CiviCRM LLC                     |
+ | at info[AT]civicrm[DOT]org. If you have questions about the        |
+ | GNU Affero General Public License or the licensing of CiviCRM,     |
+ | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
+ +--------------------------------------------------------------------+
+*}
+
+  <tr class="crm-event-eventfees-form-block-financial_type_id">
```
Comment (eileenmcnaughton) [inline templates/CRM/Contribute/Form/PaymentDetails.tpl]: Looks like quite a lot of consolidation at the tpl layer :-)

Are the css class names still appropriate?

**Comment by eileenmcnaughton** [comment]
I had a quick read through & this looks like it consolidates & improves code. I would probably focus unit tests on the search - assuming you don't want this info retrievable from the api?

File: templates/CRM/Contribute/Form/Contribution.tpl [hunk 4]
```diff
@@ -647,6 +647,29 @@ cj("#currency").on("change",function(){
...
 
+CRM.$(function($) {
+  onPaymentMethodChange();
+  $("#payment_instrument_id").on("change",function(){
+    onPaymentMethodChange();
+  });
+
+  function onPaymentMethodChange() {
+    var paymentInstrument = $('#payment_instrument_id').val();
+    if (paymentInstrument == 4) {
```
Comment (eileenmcnaughton) [inline templates/CRM/Contribute/Form/Contribution.tpl]: hard coding!

File: CRM/Member/Form/Membership.php [hunk 5]
```diff
@@ -279,10 +279,6 @@ public function setDefaultValues() {
 
     //setting default join date and receive date
     list($now, $currentTime) = CRM_Utils_Date::setDateDefaults();
-    if ($this->_action == CRM_Core_Action::ADD) {
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: how does this fit?

File: CRM/Financial/DAO/FinancialTrxn.php [hunk 6]
```diff
@@ -340,6 +365,15 @@ static function &fields() {
             'type' => 'Text',
           ) ,
         ) ,
+        'financial_trxn_credit_card_number' => array(
```
Comment (eileenmcnaughton) [inline CRM/Financial/DAO/FinancialTrxn.php]: can we call this credit_card_pan instead? we don't want to imply the number is stored

**Comment by eileenmcnaughton** [comment]
This looks thoroughly stalled & stale :-(

Can I suggest you create a new PR with the schema changes and we aim to get them merged fairly quickly?

The schema changes are the addition of the credit_card_type and the credit_card_number to the financial transaction schema. This makes sense to me as it is data specific to a transaction and there is no reason not to retain it

I am uncomfortable calling it the credit_card_number because that implies the full number. 'pan' is one industry term for the last 4 digits. I also wonder if we could come up with a more generic term that might all a reference for other payment processors (@karing @JoeMurray  any thoughts on that?)

**Comment by KarinG** [comment]
Yes to PAN (Primary Account Number) -> it is also more generic - and can apply to any Card Number -> of either a Debit or a Credit Card. We can even extend it's meaning to Account Number (ACHEFT) -> "PAN truncation" is the exact word for showing only last 4 digits and then masking/replacing the other digits with asterisks. Let's use "PAN_truncation". And perhaps go with "card_type" instead of "credit_card_type" as well?

**Comment by Edzelopez** [comment]
Rebased PR

File: CRM/Member/Form/Membership.php [hunk 7]
```diff
@@ -1488,6 +1441,9 @@ public function submit() {
       );
       $params['source'] = $formValues['source'] ? $formValues['source'] : $params['contribution_source'];
       $params['trxn_id'] = CRM_Utils_Array::value('trxn_id', $result);
+      $params['payment_instrument_id'] = 1;
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: This line needs to go - hard coding! The payment processor normally determines the default (it's stored on the processor).

**Comment by eileenmcnaughton** [comment]
I took a quick look at this - and it feels like a lot of change to make sense of.

But, my first issue is whether there is actually buy in for this change. The change basically adds a couple of fields to the civicrm_financial_trxn table: credit_card_number & credit_card_type.

@KarinG suggested calling the fields pan_truncation & card_type.

My preferred process would be
1) log a JIRA ticket which describes this change rather than linking to a closed umbrella issue
2) email the civicrm money-list directing people to the ticket and asking for people to add any input
3) assuming people agree it is a valid addition to core & can agree on a name for the fields I would suggest you put the addition of those fields (only) into a PR linked to the issue. Agreed schema-only changes are usually easy & quick to review but suffer from going stale so it's often useful to separate them out & get them merged.
4) create a new PR of the rest of the changes (probably rebased & some squashed & cleaned) referring to the new ticket.

File: CRM/Contribute/Form/Contribution.php (hunk 1 above)
Comment (pradpnayak) [inline CRM/Contribute/Form/Contribution.php]: Since buildoptions() fetches contribution status array in the form for array(value => Label).

**Comment by pradpnayak** [comment]
closing in favour of 
https://github.com/civicrm/civicrm-core/pull/9876
https://github.com/civicrm/civicrm-core/pull/9877
https://github.com/civicrm/civicrm-core/pull/9878
https://github.com/civicrm/civicrm-core/pull/9879
https://github.com/civicrm/civicrm-core/pull/9880
https://github.com/civicrm/civicrm-core/pull/9881

**Pull request closed without merging**